src/opensearch.ts

```diff
@@ -490,6 +490,73 @@ export const findBalanceByAddress =
     );
   };
 
+  export const findBalances =
+  (os: Client) =>
+  (
+    currencyIdentifier: string | null,
+    term: string,
+  ): TaskEither<ApplicationError, Result<Array<Balance>>> => {
+
+    const sort = {
+      options: [
+        {
+          sortField: "snapshotOrdinal",
+          searchDirection: SearchDirection.Before,
+        },
+      ],
+    };
+
+    // Fetch balances for specific ordinal if a valid ordinal provided
+    if (isOrdinal(term)) {
+      return pipe(
+        findAll<OpenSearchBalance>(
+          os.search(
+            getByFieldQuery<OpenSearchBalance, "snapshotOrdinal">(
+              currencyIdentifier ? OSIndex.CurrencyBalances : OSIndex.Balances,
+              "snapshotOrdinal",
+              term,
+              sort,
+              currencyIdentifier
+            )
+          ),
+          currencyIdentifier
+        ),
+        map((balances: OpenSearchBalance[]) => ({
+          data: balances.map(({ snapshotOrdinal, balance, address }) => ({ ordinal: snapshotOrdinal, balance, address })),
+          meta: {},
+        }))
+      );
+    }
```

Comment:
DRY

That part is the same as 537 - 558.

I am not sure if you are familiar with fp-ts/functional-programming but you can extract ordinal part to const and lift into `TaskEither` (`of`) then just pass it to computations.

```ts
const snapshotOrdinal = isOrdinal(term)
      ? of(term)
      : pipe(
          findSnapshot(os)("latest", currencyIdentifier),
          map((s) => s.data.ordinal)
        );

return pipe(
     snapshotOrdinal,
     chain((ordinal) =>
        findAll<OpenSearchBalance>(
          os.search(
            getByFieldQuery<OpenSearchBalance, "snapshotOrdinal">(
              currencyIdentifier ? OSIndex.CurrencyBalances : OSIndex.Balances,
              "snapshotOrdinal",
              ordinal,
              sort,
              currencyIdentifier
            )
          ),
          currencyIdentifier
        )
      ),
      map((balances: OpenSearchBalance[]) => ({ ... })
    );
```

Reply:
I felt like that could be improved but I'm not that familiar with fp-ts, thank you for explaining this and providing the code to improve the function. It's a much more elegant and clean solution :-) Changed it accordingly and also tested it locally